src/lambda_function/raw_sync/app.py

```diff
@@ -0,0 +1,518 @@
+"""
+Raw Sync Lambda
+
+This script verifies that the input and raw S3 buckets are synchronized.
+
+This is accomplished by verifying that all non-zero sized JSON in each export
+in the input S3 bucket, excepting "Manifest.json", have a corresponding object
+in the raw S3 bucket.  Because we only download the central directory, typically
+located near the end of a zip archive, verification can be done extremely quickly
+and without needing to download most of the export.
+
+If a JSON file from an export is found to not have a corresponding object in the raw bucket,
+the export is submitted to the raw Lambda (via the dispatch SNS topic) for processing.
+"""
+
+import json
+import logging
+import os
+import struct
+import zipfile
+from collections import defaultdict
+from io import BytesIO
+from typing import Optional
+
+import boto3
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+def lambda_handler(event: dict, context: dict) -> None:
+    """
+    Entrypoint for this Lambda.
+
+    Args:
+        event (dict)
+        context (dict): Information about the runtime environment and
+            the current invocation
+    """
+    s3_client = boto3.client("s3")
+    input_bucket = os.environ.get("INPUT_S3_BUCKET")
+    input_key_prefix = os.environ.get("INPUT_S3_KEY_PREFIX")
+    raw_bucket = os.environ.get("RAW_S3_BUCKET")
+    raw_key_prefix = os.environ.get("RAW_S3_KEY_PREFIX")
+    dispatch_sns_arn = os.environ.get("SNS_TOPIC_ARN")
+    main(
+        event=event,
+        s3_client=s3_client,
+        input_bucket=input_bucket,
+        input_key_prefix=input_key_prefix,
+        raw_bucket=raw_bucket,
+        raw_key_prefix=raw_key_prefix,
+        dispatch_sns_arn=dispatch_sns_arn,
+    )
+
+
+def append_s3_key(key: str, key_format: str, result: dict) -> None:
+    """
+    Organizes an S3 object key by appending it to the appropriate entry in the result dictionary
+
+    This is a helper function for `list_s3_objects`.
+
+    Args:
+        key (str): The S3 object key to process.
+        key_format (str): The format of the key, either "raw" or "input".
+        result (dict): The dictionary where keys are appended. For the "raw" format, it is a
+                       nested dictionary structured as result[data_type][cohort]. For "input",
+                       it is structured as result[cohort].
+
+    Returns:
+        None
+    """
+    if not key.endswith("/"):  # Ignore keys that represent "folders"
```

Comment:
Nit: Is there a better way to achieve this?

Reply:
Not that I know of.